### PR TITLE
Documentation changes for #1940

### PR DESCRIPTION
**Pull request by btrotta-bom**
Documentation changes associated with https://github.com/metoppv/improver/pull/1940

@benowen-bom I've incorporated your review comments from the previous PR.

### CODE REVIEW

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 1]
```diff
@@ -27,86 +26,52 @@ a suitable set of meteorological variables can be used to distinguish different
...
+threshold (note that each GBDT model itself consists of several hundred trees).
 
-Calibration of the ensemble forecast proceeds by determining the underlying distribution on
-a per realization basis and applying this to each forecast value to produce a series of
-distinct calibrated forecast values for each realization. Collectively these values form the
-so-called super-ensemble, which we subsequently sample to produce the calibrated forecast
-ensemble output.
+The input variables for the calibration are generally the NWP forecasts of the variable of interest, 
+and related variables (for example, when calibrating precipitation, a related variable would 
+be convective precipitation), along with ancillary variables such as solar time.
```
Comment (benowen-bom) [inline doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:58]: I think this is a good change, but I might suggest the following to be more focused:
```suggestion
The set of variables that feed into the GBDT models are what allows the calibration to distinguish between
different weather situations and build the appropriate probability distribution accordingly. These input
variables are typically diagnostic variables for the ensemble realization (including the variable of interest),
but can include static and dynamic ancillary variables, such as local solar time, and whole-of-ensemble
values for diagnostic variables, such as mean or standard-deviation.
```

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 2]
```diff
@@ -132,27 +97,16 @@ Model training
...
+by using only the control realization in each ensemble). 
+Each model predicts the probability that rainfall will exceed a particular threshold, 
+so the output is expected to be a number between 0 and 1. The Improver code will cap model predictions outside 
+this range so they lie between 0 and 1. (This is necessary because we are currently training using the mean 
+squared error loss, there are occasional cases where the LightGBM model predicts vlaues that are slightly more than 1 
+or less than 0.)
 
-Currently model training is done offline, using a minimum 12-month period to capture the
-full seasonal cycle.
+Currently model training is done offline.
```
Comment (benowen-bom) [inline doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:109]: Is there are reason to drop the comment about the length of the training period?

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 3]
```diff
@@ -132,27 +97,16 @@ Model training
...
-pair be as consistent as possible.
-
-For the forecast values used in training, we use the control at the shortest available
-lead-times that distinctly represent of the underlying weather types. Specifically, we
-take the earliest 24-hour period for daily accumulations, and all lead-times over the
-earliest 24-hour period for sub-daily accumulations to capture the full cycle of diurnal
-variations.
+dataframe. In general, each ensemble member yields a separate row of the dataframe (although for 
+reasons of computational efficiency it may be desirable to only use a subset of members, for example 
+by using only the control realization in each ensemble).
```
Comment (benowen-bom) [inline doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:102]: I thin k this is much better. I'd recommend dropping the brackets:
```suggestion
dataframe. In general, each ensemble member yields a separate row of the dataframe, although for 
reasons of computational efficiency it may be desirable to only use a subset of members, such as the
control realization only.
```

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 4]
```diff
@@ -132,27 +97,16 @@ Model training
...
-earliest 24-hour period for sub-daily accumulations to capture the full cycle of diurnal
-variations.
+dataframe. In general, each ensemble member yields a separate row of the dataframe (although for 
+reasons of computational efficiency it may be desirable to only use a subset of members, for example 
+by using only the control realization in each ensemble). 
+Each model predicts the probability that rainfall will exceed a particular threshold, 
+so the output is expected to be a number between 0 and 1. The Improver code will cap model predictions outside 
+this range so they lie between 0 and 1. (This is necessary because we are currently training using the mean 
+squared error loss, there are occasional cases where the LightGBM model predicts vlaues that are slightly more than 1 
+or less than 0.)
```
Comment (benowen-bom) [inline doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:107]: I think this is also a good update, but I might suggest rewording slightly:
```suggestion
As each model predicts the probability that rainfall will exceed a particular threshold, 
the output is expected to be a number between 0 and 1. However, on occasion the GBDT models may
predict values slightly outside of this range and so the model predictions are capped to back onto this
interval; such values are possible due to the current choice of loss function used in training, namely the
mean squared loss function which is akin to Brier score when working in probability space.
```

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 5]
```diff
@@ -162,38 +116,17 @@ Forecast calibration uses the trained GBDT models, along with the forecast cube
...
 
-1. Evaluate the error CDF defined over the series of error-thresholds used in model training.
-   Each exceedance probability is evaluated using the corresponding tree-model, and the feature
-   variables as inputs.
+The distributions for the individual ensemble members are formed in a two-step process:
 
-2. Interpolate the CDF to extract a series of percentile values for the error distributions.
-   The error percentiles are then added to each associated ensemble realization from the
-   forecast variable to produce a series of realisable forecast values.
+1. Evaluate CDF defined over the specified model thresholds for each ensemble member.
```
Comment (benowen-bom) [inline doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:125]: 1. Evaluate the CDF defined over the specified model thresholds for each ensemble member.

**Review by benowen-bom — COMMENTED** [review]
Thanks @btrotta-bom for making the suggested updates. I think this is fairly close, however I've provided a few more suggestions. These are focused on:
* reformatting certain paragraphs to avoid large blocks within ( )'s,
* Rewording the paragraph around input variables to generalise the why and what are used,
* some minor tweaks to sentences that I think help flow better.
Have a look and see what you think.

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 6]
```diff
@@ -27,86 +26,52 @@ a suitable set of meteorological variables can be used to distinguish different
...
-One advantage of the ECPoint approach is that the calibration is inherently non-local. As
-the calibration is done by identifying distinct weather types, the model bias and scale
-difference should be independent of any given location and time as the underlying physical
-process should be identical. Thus a grid-point can be calibrated using data from any location,
-provided the underlying weather type is the consistent. This enables effective calibration to
-be applied to areas that are typically lacking sufficient cases to calibrate against.
+Rainforests processes each ensemble member separately. The output produced from each ensemble member can be 
+considered a pseudo-ensemble which represents a 
+conditional probability distribution that describes the likelihood of observing a given outcome when the 
+realised atmospheric state is consistent with that represented in the input ensemble member forecast.
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst]: done

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 7]
```diff
@@ -27,86 +26,52 @@ a suitable set of meteorological variables can be used to distinguish different
...
-difference should be independent of any given location and time as the underlying physical
-process should be identical. Thus a grid-point can be calibrated using data from any location,
-provided the underlying weather type is the consistent. This enables effective calibration to
-be applied to areas that are typically lacking sufficient cases to calibrate against.
+Rainforests processes each ensemble member separately. The output produced from each ensemble member can be 
+considered a pseudo-ensemble which represents a 
+conditional probability distribution that describes the likelihood of observing a given outcome when the 
+realised atmospheric state is consistent with that represented in the input ensemble member forecast. 
+The predicted distributions for different ensemble members are then blended in probability space to produce 
+the final output.
```
[the remark this answers is not in the document]
Comment (btrotta-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst]: I've changed to "calibrated probability output", it's not really an ensemble.

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst (hunk 1 above)
Comment (btrotta-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst]: done

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst (hunk 2 above)
Comment (btrotta-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst]: not really, restored it now

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst (hunk 3 above)
Comment (btrotta-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst]: done

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst (hunk 5 above)
Comment (btrotta-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst]: done

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst (hunk 4 above)
Comment (btrotta-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst]: done

**Comment by btrotta-bom** [comment]
@benowen-bom Thanks for reviewing, suggested changes seem reasonable and I have implemented them.

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 8]
```diff
@@ -27,86 +26,54 @@ a suitable set of meteorological variables can be used to distinguish different
...
+different weather situations and build the appropriate probability distribution accordingly. These input
+variables are typically diagnostic variables for the ensemble realization (including the variable of interest),
+but can include static and dynamic ancillary variables, such as local solar time, and whole-of-ensemble
+values for diagnostic variables, such as mean or standard deviation.
 
-Another point of difference between RainForests and ECPoint is in the choice of error value
-underlying the error distribution. For RainForests we have chosen to use additive error in
-place of a multiplicative error (forecast error ratio in ECPoint) to allow us to calibrate
-input forecast values of zero-rainfall.
+We use LightGBM for training the models, and compile the models with Treelite for efficient prediction.
```
Comment (benowen-bom) [inline doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:62]:
```suggestion
Here we use LightGBM for training the models, and compile the models with Treelite for efficient prediction.
```

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 9]
```diff
@@ -27,86 +26,54 @@ a suitable set of meteorological variables can be used to distinguish different
...
-
-The underlying principle of the calibration methodology is essentially the same, namely using
-a set of feature variables to map onto error distributions which can be applied to the
-forecast values to produce calibrated forecast values. However, the way in which these
-distributions are constructed differs somewhat within the RainForests framework.
-
-Our approach is to use a series of GBDT models, taking the feature variables as inputs, to
-produce exceedance probability values for representative error thresholds. Collectively
-these represent the error distributions as a cumulative distribution function.
+decision tree (GBDT) models.
```
Comment (benowen-bom) [inline doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:49]:
```suggestion
```

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 10]
```diff
@@ -162,38 +119,17 @@ Forecast calibration uses the trained GBDT models, along with the forecast cube
 feature cubes. The tree-models are passed in via a model-config json which identifies
 the appropriate tree-model file for each error-threshold.
```
Comment (benowen-bom) [inline doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:120]:
```suggestion
the appropriate tree-model file for each threshold.
```

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 11]
```diff
@@ -162,38 +119,17 @@ Forecast calibration uses the trained GBDT models, along with the forecast cube
...
-   The error percentiles are then added to each associated ensemble realization from the
-   forecast variable to produce a series of realisable forecast values.
+1. Evaluate the CDF defined over the specified model thresholds for each ensemble member.
+Each threshold exceedance probability is evaluated using the corresponding
+decision-tree model.
 
-Collectively these series form the calibrated super-ensemble which is obtained by collapsing
-the two realization dimensions into one. This is then sampled to provide the calibrated
-ensemble forecast.
+2. Interpolate each ensemble member distribution to the output thresholds.
```
Comment (benowen-bom) [inline doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:132]: The formatting in the markdown seems a bit off. Might be worth double checking what's causing the misalignment (maybe the extra line)?

**Review by benowen-bom — DISMISSED** [review]
Thanks @btrotta-bom. I've had one last review, but this time reading the markdown directly rather than this diff. I've got a few minor suggestions to polish this off; the two I strongly recommend are the formatting of the list and the instance of `error-threshold`. With those two changes, I'm happy to go ahead and approve.

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst (hunk 9 above)
Comment (btrotta-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:49]: what did you mean here? it seems the suggested change just deleted the end of the sentence...

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst (hunk 10 above)
Comment (btrotta-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst]: done

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst (hunk 8 above)
Comment (btrotta-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst]: done

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst (hunk 11 above)
Comment (btrotta-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst]: fixed, it was the extra line

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst (hunk 9 above)
Comment (benowen-bom) [reply doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:49]: Sorry, I had another suggestion above this that hasn't seemed to have made it in. I'll add that comment in too... given this straddles both additions and removals, I was trying to frame the suggestions so that they could easily slot in (but I obviously failed).

File: doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst [hunk 12]
```diff
@@ -27,86 +26,54 @@ a suitable set of meteorological variables can be used to distinguish different
...
+sufficient cases to calibrate against.
+
 
 ****************************
 The RainForests method
 ****************************
 
-RainForests is an adaptation of the ECPoint method that seeks to use machine learning based
+RainForests is an adaptation of the ECPoint method that uses machine learning based
 tree models to replace the manually constructed tree model of ECPoint. Here we use gradient-boosted
```
Comment (benowen-bom) [inline doc/source/extended_documentation/calibration/rainforests_calibration/rainforests_calibration.rst:48]:
```suggestion
RainForests uses machine learning based tree models, namely gradient-boosted decision tree
(GBDT) models to replace the manually constructed decision tree model of ECPoint.
```

**Review by benowen-bom — APPROVED** [review]
Great, thanks for making those final changes. I'm happy with the updates to the documentation now.
